### ISSUE TITLE
Prematurely Plays - Play a Sound When a Download is Finished

### ISSUE THREAD

**Issue by dbareis** [BiglySoftware/BiglyBT#2461]
I have the "Re-Check pieces when downloading is done" option ticked so the download hasn't completed until the (slow) recheck occurs, yet the sound plays before it's started yet alone completed.

---
Java 1.8.0_202 (64 bit)
  Oracle Corporation
c:\program files\biglybt\jre

SWT v4942r22, win32, zoom=100, dpi=96
Windows 10 v10.0, amd64 (64 bit)
B2.9.0.1_B35/4 az2

**Comment by parg** [comment]
That is how it works, it isn't "play a sound when a download is rechecked".  Of course it would be possible to add that option, do you really need it? The recheck on completion really is "just in case" - each piece of a download is checked for correctness during the download process so the likelyhood of an error being found during the recheck is very small.

**Comment by dbareis** [comment]
I don't think "how it works" is an argument (except that Microsoft use it to convert "bugs" into "features"), You have to ask yourself what is the intention of the alert?  Is it to alert the user that the download is ready?  If so it doesn't do that now...

I don't think there needs to be an option the situation is clear if I have the recheck option on the "download" hasn't completed until the recheck is complete.

It is irrelevant how unlikely the need for a recheck is. The feature exists an unlikely is not 0%.  "Just in Case" is why you have insured your house.

**Comment by parg** [comment]
Thanks for the insight

**Comment by ferdnyc** [comment]
For a far less antagonistic attempt at this, some thoughts:

@parg
> The recheck on completion really is "just in case" - each piece of a download is checked for correctness during the download process so the likelyhood of an error being found during the recheck is very small.

Fair, but it's also a default-enabled option, so people are pretty likely to have it turned on.

I was originally going to ask, "what happens with completion moving, does that wait for the recheck to complete?" But then I looked, and the answer is, "Well, it depends on the settings." Specifically, the child checkbox here:

![Screenshot from 2023-01-09 13-18-23](https://user-images.githubusercontent.com/538020/211380840-ff423c59-0ad3-4145-9665-7ed92e8da937.png)

If, in theory, that checkbox is _also checked_, then the file won't be available in its ultimate destination until the recheck, and then the move, both complete. So, that's the point where I think "download finished" and "torrent completed" are different enough that both may warrant possible audio alerts of their own.

@dbareis 
> You have to ask yourself what is the intention of the alert?

Well, exactly! And I don't think there's only one answer to that question. The simple fact is, different users may have _different_ intentions for that alert.

If you're looking to use it as a "torrent is complete and ready for use" alert, then I agree that, under some circumstances, it may not represent that.

But if the user's intention is to be alerted when BiglyBT has stopped consuming internet bandwidth (say, because they want to stream something, or browse the web without any torrent-induced slowness), then the existing alert is perfectly suited. The re-check, move, and other post-download activities require minimal network bandwidth, if any at all. So, I'd hate to see it just change to the way you want it, and _no longer_ suit the needs of **those** users as a result.

I don't think anyone can say that playing the sound for either event is _unequivocally_ the correct time, and that the other is _unequivocally_ the wrong time to play it. Which is why I think it probably makes sense to offer both options. Just like there's a separate popup alert setting for when a download is checked:

![image](https://user-images.githubusercontent.com/538020/211383150-8c573d13-d4b2-444c-81f2-c3ee813b3a02.png)

(Though I'm not suggesting the new sound alert should be "Play a sound when a download is checked". I don't think anyone is really looking for that. Popup alerts are useful _during_ a process, and should often be shown when it begins. Whereas audio alerts mostly signify the _end_ of some process. They're fundamentally different things.)

But something like "Play a sound when a torrent is completed" would probably be useful to some users. Where "completed" implies:
* re-checking done (if enabled)
* completion copy/move done (if enabled)
* basically all activity related to the torrent (other than seeding, of course) done
 
Heck, if post-download automatic transcoding is an option (is it, or is transcoding only on-the-fly/on-demand in the media server?) then I'd say it should even wait until after **that's** finished.

**Comment by parg** [comment]
Maybe I'll add a Tag Constraint of "isMoving" and "execute on assign" action of "play a sound". 

That way people can set up their own audio notifications based on whatever they want.

Although that wouldn't support your transcoding example...

**Comment by parg** [comment]
With the next beta you can setup a tag with a constraint of (for example)

`isComplete() && !isChecking() && !isMoving()`

with an "execute on assign" script value of (for example)

`plugin( simpleapi, "method=playsound&file=c:\temp\taunt.wav" )`

to play a sound when the download is completed and not checking or moving.

**Comment by dbareis** [comment]
I'm using the beta, I looked at "alerts" & "Tags" neither look like I can do what you say....

**Comment by parg** [comment]
works for me

**Comment by parg** [comment]
In case you missed it, my earlier post (2 weeks ago) tells you how to set it up. It is nothing to do with "alerts". There is a new feature that lets you configure a Tag so that when a download is assigned to it a sound is played. You can select your own sound or use the default one. 

1) go to the View menu and select "Tags Overview" to see all of your existing Tags
2) create a new Tag using the "Add Tag.." button top right of this view - call it, say, PlaySound
3) find the PlaySound Tag in the list, right-click it, select the "Execute On Assign" menu item and then the "Script..." sub-menu item
4) Enter `plugin( simpleapi, "method=playsound" )` into the script box
5) So far the Tag is configured so that the default sound will be played whenever a download is assigned to it. As it is you would need to manually assign downloads to it as it has no "constraint" setup to automatically assign things to it. You can test it is working by manually assigning a download to it - for example by right-clicking on a download in your Library view, selecting the "Organise" tab and then the Tags menu. For your purpose you want downloads to be automatically assigned to the Tag when they are complete, not checking and not moving. To do this you set up a constraint on the Tag:
6) Go to Tags Overview and select PlaySound in the list of Tags. In the bottom area select the "Settings" tab and scroll down to the Constraint box. Enter `isComplete() && !isChecking() && !isMoving()` into the box and hit Save

**Comment by dbareis** [comment]
Thanks for the detailed explanation. It works except for the sound, as far as I can see there is no "simpleapi" plugin installed and when I try to "get" them there, "simpleapi" is not listed there either.

**Comment by parg** [comment]
That plugin is built in, you don't need to install it.

**Comment by dbareis** [comment]
In that case where are the configuration for it.  It currently doesn't produce a sound and in anycase I want to use the sound I've been using.

**Comment by parg** [comment]
Perhaps you haven't configured the "execute on assign" correctly. It plays the default sound when set up as above. Get that working and then use the extended form to specify your own sound

`plugin( simpleapi, "method=playsound&file=c:\temp\taunt.wav" )`

**Comment by dbareis** [comment]
I'm not sure, maybe it was too subtle or high pitched, my sound works, thanks.
